**Problem:**

You are given the following dataset:
1. **Audible Data** : https://1drv.ms/u/s!AiqdXCxPTydhoog8ckLN-6Cw55fzIg?e=EWgZ5d

Your task is to:
- Find the problems with the datasets.
- Define the Data Quality Dimensions.
- Try to clean the datasets.

In [2]:
import numpy as np
import pandas as pd

In [4]:
audible = pd.read_excel("/content/audible_uncleaned.xlsx",sheet_name = "in")

In [6]:
audible

name                     author  \
0                      Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                                The Burning Maze      Writtenby:RickRiordan   
2                                    The Deep End       Writtenby:JeffKinney   
3                            Daughter of the Deep      Writtenby:RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   
...                                           ...                        ...   
87484                   Last Days of the Bus Club     Writtenby:ChrisStewart   
87485                                    The Alps    Writtenby:StephenO'Shea   
87486                        The Innocents Abroad        Writtenby:MarkTwain   
87487                       A Sentimental Journey   Writtenby:LaurenceSterne   
87488                                      Havana    Writtenby:MarkKurlansky   

                        narrator                time          releasedate  \
0          Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1       Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2          Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3      Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4      Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   
...                          ...                 ...                  ...   
87484    Narratedby:ChrisStewart   7 hrs and 34 mins  2017-09-03 00:00:00   
87485      Narratedby:RobertFass   10 hrs and 7 mins             21-02-17   
87486       Narratedby:FloGibson   19 hrs and 4 mins             30-12-16   
87487     Narratedby:AntonLesser    4 hrs and 8 mins             23-02-11   
87488     Narratedby:FleetCooper     6 hrs and 1 min  2017-07-03 00:00:00   

      language                          stars price  
0      English     5 out of 5 stars34 ratings   468  
1      English   4.5 out of 5 stars41 ratings   820  
2      English   4.5 out of 5 stars38 ratings   410  
3      English   4.5 out of 5 stars12 ratings   615  
4      English  4.5 out of 5 stars181 ratings   820  
...        ...                            ...   ...  
87484  English                  Not rated yet   596  
87485  English                  Not rated yet   820  
87486  English                  Not rated yet   938  
87487  English                  Not rated yet   680  
87488  English                  Not rated yet   569  

[87489 rows x 8 columns]

### 1. Write a summary for your data

This the dataset contains about 87489 audibles with there respective writers and narrators. It is also contains the audibles language and also with the time of it. It is also contains the rating and reviews of the respective audibles and with there price. Data also contains the date of the release.


### 2. Write Column descriptions

- `name` -> the name of the corresponding Audible
- `author` - > Name of the author for corresponding Audible
- `narrator` -> Name of the narrator for corresponding Audible
- `time` ->duration of corresponding Audible
- `release date` ->date of release of corresponding Audible
- `language` -> language of corresponding Audible
- `stars` -> rating and reviews of corresponding Audible
- `price`-> price of corresponding Audible






### Issues with the dataset

1. Dirty Data
  - author column contains "written by" text before the author name `Validity`
  - narrator column contains "narrated by" text before the narrator name `Validity`
  - in narrator column some values are "Anonymous" and "anonymous".`Consistency`
  - time column have both mins and hrs -> should be only mins.`Consistency`
  - in data columns some values contain like '01/01/2002' and some are '01-01-2002' `consistency`
  - price column contains "," `validity`
  - time columns does not need "mins" text after converting into mins `validity`
  - stars column contains extra text "5 out of 5 stars" and "3 ratings" `validity`
  - in language column some languages are Captalized and some are lowercase `consis`
  - incorrect dtype for time `Validity`
  - incorrect dtype for relesedate `Validity`
  - incorrect dtype for price `Validity`
  - 'not' in rating col and 'rated' in total rating col treated as nan `Validity`
  - incorrect dtype for rating `Validity`
  - incorrect dtype for total rating `Validity`


2. Messy Data
  - both rate and total number of ratings are in same column 'stars'only.


In [5]:
audible_df = audible.copy()

In [6]:
audible_df.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language                          stars price  
0  English     5 out of 5 stars34 ratings   468  
1  English   4.5 out of 5 stars41 ratings   820  
2  English   4.5 out of 5 stars38 ratings   410  
3  English   4.5 out of 5 stars12 ratings   615  
4  English  4.5 out of 5 stars181 ratings   820

### Describe
- both rate and total number of ratings are in same column 'stars'only.

In [7]:
# code
audible_df['rating'] = audible_df['stars'].str.split(" ").str.get(0)

In [8]:
# code
audible_df['Total rating'] = audible_df['stars'].str.split(" ").str.get(-2).str.split("s").str.get(-1)

In [9]:
# Test
audible_df.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language                          stars price rating Total rating  
0  English     5 out of 5 stars34 ratings   468      5           34  
1  English   4.5 out of 5 stars41 ratings   820    4.5           41  
2  English   4.5 out of 5 stars38 ratings   410    4.5           38  
3  English   4.5 out of 5 stars12 ratings   615    4.5           12  
4  English  4.5 out of 5 stars181 ratings   820    4.5          181

In [10]:
audible_df.drop(columns = 'stars',inplace = True)

In [11]:
# code
audible_df['author']  = audible_df['author'].str.split(":").str.get(1)

In [12]:
# test
audible_df.head()

name           author  \
0                  Geronimo Stilton #11 & #12  GeronimoStilton   
1                            The Burning Maze      RickRiordan   
2                                The Deep End       JeffKinney   
3                        Daughter of the Deep      RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language price rating Total rating  
0  English   468      5           34  
1  English   820    4.5           41  
2  English   410    4.5           38  
3  English   615    4.5           12  
4  English   820    4.5          181

In [13]:
# code
audible_df['narrator']  = audible_df['narrator'].str.split(":").str.get(1)

In [14]:
# test
audible_df.head()

name           author  \
0                  Geronimo Stilton #11 & #12  GeronimoStilton   
1                            The Burning Maze      RickRiordan   
2                                The Deep End       JeffKinney   
3                        Daughter of the Deep      RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      RickRiordan   

         narrator                time          releasedate language price  \
0      BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00  English   468   
1   RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00  English   820   
2      DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00  English   410   
3  SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00  English   615   
4  JesseBernstein              10 hrs  2010-01-13 00:00:00  English   820   

  rating Total rating  
0      5           34  
1    4.5           41  
2    4.5           38  
3    4.5           12  
4    4.5          181

In [15]:
# code
audible_df['price'] = audible_df['price'].replace(",","")

In [16]:
#test
audible_df.sample(5)

name                      author  \
10029            Dragonblood: The Missing Fang  MichaelDahl,FedericoPiatti   
55423                      Woman at Point Zero              NawalelSaadawi   
58405  P.S.: Was ist Lust und was ist Pflicht?                MichaelVoigt   
45422                           Blut und Feuer                  WardaMoram   
44565                              The Avenger                DavidRobbins   

            narrator                time          releasedate language price  \
10029      anonymous              6 mins             16-03-22  English   105   
55423     LauraHanna   3 hrs and 48 mins             16-06-16  English   569   
58405   VolkerBrandt  12 hrs and 27 mins             23-11-16   german   669   
45422  MatthiasLühn  11 hrs and 11 mins             26-10-21   german   334   
44565    RustyNelson   7 hrs and 26 mins  2022-05-04 00:00:00  English   234   

      rating Total rating  
10029    Not        rated  
55423    4.5            4  
58405    Not        rated  
45422    Not        rated  
44565    Not        rated

In [17]:
#code
audible_df["releasedate"] = pd.to_datetime(audible_df["releasedate"])

In [18]:
#test
audible_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   name          87489 non-null  object        
 1   author        87489 non-null  object        
 2   narrator      87489 non-null  object        
 3   time          87489 non-null  object        
 4   releasedate   87489 non-null  datetime64[ns]
 5   language      87489 non-null  object        
 6   price         87489 non-null  object        
 7   rating        87489 non-null  object        
 8   Total rating  87489 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 6.0+ MB


In [19]:
# test
audible_df.head()

name           author  \
0                  Geronimo Stilton #11 & #12  GeronimoStilton   
1                            The Burning Maze      RickRiordan   
2                                The Deep End       JeffKinney   
3                        Daughter of the Deep      RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      RickRiordan   

         narrator                time releasedate language price rating  \
0      BillLobely   2 hrs and 20 mins  2008-04-08  English   468      5   
1   RobbieDaymond   13 hrs and 8 mins  2018-01-05  English   820    4.5   
2      DanRussell    2 hrs and 3 mins  2020-06-11  English   410    4.5   
3  SoneelaNankani  11 hrs and 16 mins  2021-05-10  English   615    4.5   
4  JesseBernstein              10 hrs  2010-01-13  English   820    4.5   

  Total rating  
0           34  
1           41  
2           38  
3           12  
4          181

In [20]:
#code
audible_df['rating'] = audible_df['rating'].replace("np.nan",np.nan)

In [21]:
#test
audible_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   name          87489 non-null  object        
 1   author        87489 non-null  object        
 2   narrator      87489 non-null  object        
 3   time          87489 non-null  object        
 4   releasedate   87489 non-null  datetime64[ns]
 5   language      87489 non-null  object        
 6   price         87489 non-null  object        
 7   rating        87489 non-null  object        
 8   Total rating  87489 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 6.0+ MB


In [22]:
audible_df

name           author  \
0                      Geronimo Stilton #11 & #12  GeronimoStilton   
1                                The Burning Maze      RickRiordan   
2                                    The Deep End       JeffKinney   
3                            Daughter of the Deep      RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1      RickRiordan   
...                                           ...              ...   
87484                   Last Days of the Bus Club     ChrisStewart   
87485                                    The Alps    StephenO'Shea   
87486                        The Innocents Abroad        MarkTwain   
87487                       A Sentimental Journey   LaurenceSterne   
87488                                      Havana    MarkKurlansky   

             narrator                time releasedate language price rating  \
0          BillLobely   2 hrs and 20 mins  2008-04-08  English   468      5   
1       RobbieDaymond   13 hrs and 8 mins  2018-01-05  English   820    4.5   
2          DanRussell    2 hrs and 3 mins  2020-06-11  English   410    4.5   
3      SoneelaNankani  11 hrs and 16 mins  2021-05-10  English   615    4.5   
4      JesseBernstein              10 hrs  2010-01-13  English   820    4.5   
...               ...                 ...         ...      ...   ...    ...   
87484    ChrisStewart   7 hrs and 34 mins  2017-09-03  English   596    Not   
87485      RobertFass   10 hrs and 7 mins  2017-02-21  English   820    Not   
87486       FloGibson   19 hrs and 4 mins  2016-12-30  English   938    Not   
87487     AntonLesser    4 hrs and 8 mins  2011-02-23  English   680    Not   
87488     FleetCooper     6 hrs and 1 min  2017-07-03  English   569    Not   

      Total rating  
0               34  
1               41  
2               38  
3               12  
4              181  
...            ...  
87484        rated  
87485        rated  
87486        rated  
87487        rated  
87488        rated  

[87489 rows x 9 columns]

In [23]:
#code
audible_df['Toatal rating'] = audible_df['Total rating'].replace("rated",np.nan)


In [24]:
#test
audible_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   name           87489 non-null  object        
 1   author         87489 non-null  object        
 2   narrator       87489 non-null  object        
 3   time           87489 non-null  object        
 4   releasedate    87489 non-null  datetime64[ns]
 5   language       87489 non-null  object        
 6   price          87489 non-null  object        
 7   rating         87489 non-null  object        
 8   Total rating   87489 non-null  object        
 9   Toatal rating  15072 non-null  object        
dtypes: datetime64[ns](1), object(9)
memory usage: 6.7+ MB


In [25]:
import pandas as pd

# Assuming you have your durations in a list

# Create a DataFrame from the durations list
df = pd.DataFrame({'duration': audible_df['time']})

# Extract hours and minutes using regular expressions and convert them to numeric values
df['hours'] = df['duration'].str.extract('(\d+) hrs').astype(float)
df['minutes'] = df['duration'].str.extract('(\d+) mins').astype(float)

# Fill missing values in 'hours' and 'minutes' columns
df['hours'].fillna(0, inplace=True)
df['minutes'].fillna(0, inplace=True)

# Calculate the total time in minutes
df['total_minutes'] = df['hours'] * 60 + df['minutes']

# Convert the total time in minutes to proper time format
df['time'] = pd.to_datetime(df['total_minutes'], unit='m').dt.strftime('%H:%M')

# Print the resulting DataFrame
print(df)


                 duration  hours  minutes  total_minutes   time
0       2 hrs and 20 mins    2.0     20.0          140.0  02:20
1       13 hrs and 8 mins   13.0      8.0          788.0  13:08
2        2 hrs and 3 mins    2.0      3.0          123.0  02:03
3      11 hrs and 16 mins   11.0     16.0          676.0  11:16
4                  10 hrs   10.0      0.0          600.0  10:00
...                   ...    ...      ...            ...    ...
87484   7 hrs and 34 mins    7.0     34.0          454.0  07:34
87485   10 hrs and 7 mins   10.0      7.0          607.0  10:07
87486   19 hrs and 4 mins   19.0      4.0         1144.0  19:04
87487    4 hrs and 8 mins    4.0      8.0          248.0  04:08
87488     6 hrs and 1 min    6.0      0.0          360.0  06:00

[87489 rows x 5 columns]


In [26]:
audible_df['time'] = df['time']

In [27]:
#test
audible_df

name           author  \
0                      Geronimo Stilton #11 & #12  GeronimoStilton   
1                                The Burning Maze      RickRiordan   
2                                    The Deep End       JeffKinney   
3                            Daughter of the Deep      RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1      RickRiordan   
...                                           ...              ...   
87484                   Last Days of the Bus Club     ChrisStewart   
87485                                    The Alps    StephenO'Shea   
87486                        The Innocents Abroad        MarkTwain   
87487                       A Sentimental Journey   LaurenceSterne   
87488                                      Havana    MarkKurlansky   

             narrator   time releasedate language price rating Total rating  \
0          BillLobely  02:20  2008-04-08  English   468      5           34   
1       RobbieDaymond  13:08  2018-01-05  English   820    4.5           41   
2          DanRussell  02:03  2020-06-11  English   410    4.5           38   
3      SoneelaNankani  11:16  2021-05-10  English   615    4.5           12   
4      JesseBernstein  10:00  2010-01-13  English   820    4.5          181   
...               ...    ...         ...      ...   ...    ...          ...   
87484    ChrisStewart  07:34  2017-09-03  English   596    Not        rated   
87485      RobertFass  10:07  2017-02-21  English   820    Not        rated   
87486       FloGibson  19:04  2016-12-30  English   938    Not        rated   
87487     AntonLesser  04:08  2011-02-23  English   680    Not        rated   
87488     FleetCooper  06:00  2017-07-03  English   569    Not        rated   

      Toatal rating  
0                34  
1                41  
2                38  
3                12  
4               181  
...             ...  
87484           NaN  
87485           NaN  
87486           NaN  
87487           NaN  
87488           NaN  

[87489 rows x 10 columns]

In [28]:
#code
audible_df['author'] = audible_df['author'].replace("anonymous","Anonymous")


In [29]:
# test
audible_df[audible_df['author'] == "anonymous"].shape[0]


0

In [30]:
#code
audible_df['narrator'] = audible_df['narrator'].replace("anonymous","Anonymous")


In [31]:
#test
audible_df[audible_df['narrator'] == "anonymous"].shape[0]


0

In [32]:
#code
audible_df["language"] = audible_df["language"].str.capitalize()

In [33]:
#test
audible_df["language"].unique()

array(['English', 'Hindi', 'Spanish', 'German', 'French', 'Catalan',
       'Swedish', 'Italian', 'Danish', 'Finnish', 'Dutch', 'Hebrew',
       'Russian', 'Polish', 'Galician', 'Afrikaans', 'Icelandic',
       'Romanian', 'Japanese', 'Tamil', 'Portuguese', 'Urdu', 'Hungarian',
       'Czech', 'Bulgarian', 'Mandarin_chinese', 'Basque', 'Korean',
       'Arabic', 'Greek', 'Turkish', 'Ukrainian', 'Slovene', 'Norwegian',
       'Telugu', 'Lithuanian'], dtype=object)

In [35]:
audible_df.to_excel('Audible_cleaned.xlsx')